### PR TITLE
Add web API support for flashing

### PR DESCRIPTION
**Pull request by haslinghuis**
- Enabled flashing from serial connection to jump into DFU mode
- Fixes exit DFU button

KNOWN ISSUES
- Need help with double permission dialog for DFU

FOLLOW UP
- `Flash on connect` and `no reboot sequence` have not been implemented or tested - as Porthandler has been changed.

### CODE REVIEW

File: src/js/port_handler.js [hunk 1]
```diff
@@ -1,15 +1,7 @@
...
-import { i18n } from "./localization";
 import { get as getConfig } from "./ConfigStorage";
-import { isWeb } from "./utils/isWeb";
-import { usbDevices } from "./usb_devices";
-import { serialShim } from "./serial_shim.js";
-import { usbShim } from "./usb_shim.js";
 import { EventBus } from "../components/eventBus";
-
-const serial = serialShim();
-const usb = usbShim();
```
Comment (chmelevskij) [inline src/js/port_handler.js:12]: 👍 good call, we don't need this anymore anyways

File: src/js/protocols/webstm32.js [hunk 2]
```diff
@@ -0,0 +1,941 @@
...
+        TABS.firmware_flasher.flashingMessage(null, TABS.firmware_flasher.FLASH_MESSAGE_TYPES.NEUTRAL)
+            .flashProgress(0);
+
+        // lock some UI elements TODO needs rework
+        $('select[name="release"]').prop('disabled', true);
+
+        // TODO: update to use web serial / USB API
+        serial.onReceive.addListener((info) => {
+            this.read(info);
+        });
```
Comment (chmelevskij) [inline src/js/protocols/webstm32.js:332]: I don't think we even have `onReceive` in web serial implementation. We are using standard event target there.

File: src/js/msp/MSPConnector.js [hunk 3]
```diff
@@ -6,85 +6,105 @@ import MSP from "../msp";
...
+                    console.log('Connected');
 
-            serial.onReceive.addListener(read_serial);
+                    this.onConnectCallback();
+                });
+            } else {
+                gui_log(i18n.getMessage('serialPortOpenFail'));
+                this.onFailureCallback();
+            }
+        };
```
Comment (chmelevskij) [inline src/js/msp/MSPConnector.js:79]: A bit of a nitpick, but why are we attaching method inside of a method here? Let's add it just as any other method at the top of the class and then just bind it to event listeners.

File: src/js/msp/MSPConnector.js [hunk 4]
```diff
@@ -6,85 +6,105 @@ import MSP from "../msp";
...
 
-            const mspHelper = new MspHelper();
-            MSP.listen(mspHelper.process_data.bind(mspHelper));
+        this.disconnectHandler = (detail) => {
+            console.log('Disconnected', detail);
+            MSP.clearListeners();
+            MSP.disconnect_cleanup();
+        };
```
Comment (chmelevskij) [inline src/js/msp/MSPConnector.js:85]: Similar, this should be another method in the class instead of `this.bind`

File: src/js/protocols/webstm32.js [hunk 5]
```diff
@@ -0,0 +1,941 @@
...
+                    serial.connect(this.port, { bitrate: this.baud, parityBit: 'even', stopBits: 'one' }, function (openInfo) {
+                        if (openInfo) {
+                            this.initialize();
+                        } else {
+                            GUI.connect_lock = false;
+                            gui_log(i18n.getMessage('serialPortOpenFail'));
+                        }
+                    });
+                }
+            };
```
Comment (chmelevskij) [inline src/js/protocols/webstm32.js:137]: Stylistic change, but could we move this outside of the class Or create a method instead?

It just becomes very hard to follow when we define many nested functions and then use them in the same method.

File: src/js/protocols/webstm32.js [hunk 6]
```diff
@@ -0,0 +1,941 @@
...
+                    });
+                }
+            }
+
+            const onTimeoutHandler = function () {
+                GUI.connect_lock = false;
+                console.log('Looking for capabilities via MSP failed');
+
+                TABS.firmware_flasher.flashingMessage(i18n.getMessage('stm32RebootingToBootloaderFailed'), TABS.firmware_flasher.FLASH_MESSAGE_TYPES.INVALID);
+            };
```
Comment (chmelevskij) [inline src/js/protocols/webstm32.js:294]: this could just be function at the top level of module, this doesn't rely on any class internal state.

File: src/js/protocols/webstm32.js [hunk 7]
```diff
@@ -0,0 +1,941 @@
...
+
+            const onFailureHandler = function () {
+                GUI.connect_lock = false;
+                TABS.firmware_flasher.refresh();
+            };
```
Comment (chmelevskij) [inline src/js/protocols/webstm32.js:299]: Similar to the above, can be just a helper function in the same file instead of being inline.

File: src/js/protocols/webstm32.js (hunk 2 above)
Comment (haslinghuis) [reply src/js/protocols/webstm32.js]: Yes that part is still open. Fixed anyway.

File: src/js/msp/MSPConnector.js (hunk 3 above)
Comment (haslinghuis) [reply src/js/msp/MSPConnector.js]: Fixed.

File: src/js/msp/MSPConnector.js (hunk 4 above)
Comment (haslinghuis) [reply src/js/msp/MSPConnector.js]: Done.

File: src/js/protocols/webstm32.js (hunk 5 above)
Comment (haslinghuis) [reply src/js/protocols/webstm32.js]: Created a method

File: src/js/protocols/webstm32.js (hunk 6 above)
Comment (haslinghuis) [reply src/js/protocols/webstm32.js]: Done.

File: src/js/protocols/webstm32.js (hunk 7 above)
Comment (haslinghuis) [reply src/js/protocols/webstm32.js]: Thanks for review.

Done.

Due to second refactor found an issue with MSP still listening - now fixed.

**Comment by haslinghuis** [comment]
Rebased.

**Comment by nerdCopter** [comment]
`08e3c440` local `yarn vite`:
- clicking flash popped up the permissions dialog :)
- but then did nothing else.
image

**Comment by haslinghuis** [comment]
@nerdcopter - Flashed more then a 100 times. On AIRBOTSUPERF4 works too:

image

image

**Comment by haslinghuis** [comment]
Let me rebase

**Comment by nerdCopter** [comment]
`eb064202` yarn dev:
- something different for me. tried STM405 with same results as well :(
image

**Comment by haslinghuis** [comment]
Checked AOCODARCF405V3 too. No problem :exploding_head:

**Comment by nerdCopter** [comment]
`47c3148c` double DDFU permissions on this one, but it is working. (did not have double dialog prior.)

**Review by nerdCopter — APPROVED** [review]
- approving with known caveat of double-dialog to be fixed long-term.
